File: 0383-ransom-note-赎金信/0383.cpp
class Solution {
    inline bool cut(string & magazine, char k) {
        for (int i = 0; i != magazine.size(); ++i) {
            if (magazine[i] == k) {
                magazine.erase(i, 1);
                return true;
            }
        }
        return false;
    }
public:
    bool canConstruct(string ransomNote, string magazine) {
        for (auto &k : ransomNote) {
            bool isExist = cut(magazine, k);
            if (!isExist) return false;
        }
        return true;
    }
};

File: 0383-ransom-note-赎金信/0383.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character,Integer> map = new HashMap<>();
        for (int i = 0; i != magazine.length(); ++i) {
            if (map.containsKey(magazine.charAt(i))) {
                Integer v = map.get(magazine.charAt(i)) + 1;
                map.put(magazine.charAt(i), v);
            }  else {
                map.put(magazine.charAt(i), 1);
            }
        }
        
        for (int i = 0; i != ransomNote.length(); ++i) {
            Character k = ransomNote.charAt(i);
            if (map.containsKey(k)) { 
                Integer v = map.get(k) - 1;
                map.put(k, v);
                if (map.get(k) < 0) {
                    return false;
                }
            } else {
                return false;
            }
        }
        return true;
    }
}